src/Util/HTTPSig.php

```diff
@@ -0,0 +1,400 @@
+
+			if ($digest[0] === 'SHA-256') {
+				$hashalg = 'sha256';
+			}
+			if ($digest[0] === 'SHA-512') {
+				$hashalg = 'sha512';
+			}
+
+			// The explode operation will have stripped the '=' padding, so compare against unpadded base64.
+			if (rtrim(base64_encode(hash($hashalg, $body, true)), '=') === $digest[1]) {
+				$result['content_valid'] = true;
+			}
+		}
+
+		logger('Content_Valid: ' . $result['content_valid']);
+
+		return $result;
+	}
+
+	/**
+	 * Fetch the public key for Activity Pub contact.
+	 * 
+	 * @param string|int The identifier (contact addr or contact ID).
+	 * @return string|boolean The public key or false on failure.
+	 */
+	private static function getActivitypubKey($id)
+	{
+		if (strpos($id, 'acct:') === 0) {
+			$x = dba::selectFirst('contact', ['pubkey'], ['uid' => 0, 'addr' => str_replace('acct:', '', $id)]);
+		} else {
+			$x = dba::selectFirst('contact', ['pubkey'], ['id' => $id, 'network' => 'activitypub']);
+		}
+
+		if (DBM::is_result($x)) {
+			return $x['pubkey'];
+		}
+
+		if(function_exists('as_fetch')) {
+			$r = as_fetch($id);
+		}
+
+		if ($r) {
+			$j = json_decode($r, true);
+
+			if (array_key_exists('publicKey', $j) && array_key_exists('publicKeyPem', $j['publicKey'])) {
+				if ((array_key_exists('id', $j['publicKey']) && $j['publicKey']['id'] !== $id) && $j['id'] !== $id) {
+					return false;
+				}
+
+				return $j['publicKey']['publicKeyPem'];
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * @brief
+	 *
+	 * @param string  $request
+	 * @param array   $head
+	 * @param string  $prvkey
+	 * @param string  $keyid (optional, default 'Key')
+	 * @param boolean $send_headers (optional, default false)
+	 *   If set send a HTTP header
+	 * @param boolean $auth (optional, default false)
+	 * @param string  $alg (optional, default 'sha256')
+	 * @param string  $crypt_key (optional, default null)
+	 * @param string  $crypt_algo (optional, default 'aes256ctr')
+	 * 
+	 * @return array
+	 */
+	public static function createSig($request, $head, $prvkey, $keyid = 'Key', $send_headers = false, $auth = false, $alg = 'sha256', $crypt_key = null, $crypt_algo = 'aes256ctr')
+	{
+		$return_headers = [];
+
+		if ($alg === 'sha256') {
+			$algorithm = 'rsa-sha256';
+		}
+
+		if ($alg === 'sha512') {
+			$algorithm = 'rsa-sha512';
+		}
+
+		$x = self::sign($request, $head, $prvkey, $alg);
+
+		$headerval = 'keyId="' . $keyid . '",algorithm="' . $algorithm
+			. '",headers="' . $x['headers'] . '",signature="' . $x['signature'] . '"';
+
+		if ($crypt_key) {
+			$x = Crypto::encapsulate($headerval, $crypt_key, $crypt_algo);
+			$headerval = 'iv="' . $x['iv'] . '",key="' . $x['key'] . '",alg="' . $x['alg'] . '",data="' . $x['data'] . '"';
+		}
+
+		if ($auth) {
+			$sighead = 'Authorization: Signature ' . $headerval;
+		} else {
+			$sighead = 'Signature: ' . $headerval;
+		}
+
+		if ($head) {
+			foreach ($head as $k => $v) {
+				if ($send_headers) {
+					header($k . ': ' . $v);
```

Comment:
Nope again for the same reason as before. `createSig` shouldn't set headers.

Reply:
If this is the AP part, we could leave this whole block as a comment.

Reply:
It doesn't seem to be specific to ActivityPub.